Refactor C++ language extensions and C++ support (#3673)
ROCm-Developer-Tools/HIP: docs/how-to/hip_cpp_language_extensions.rst

================================================================================

HIP introduces three different function qualifiers to mark functions for
execution on the device or the host, and also adds new qualifiers to control
inlining of functions.

.. _host_attr:

__host__
--------------------------------------------------------------------------------

The ``__host__`` qualifier is used to specify functions for execution
on the host. This qualifier is implicitly defined for any function where no
``__host__``, ``__device__`` or ``__global__`` qualifier is added, in order to
not break compatibility with existing C++ functions.

You can't combine ``__host__`` with ``__global__``.

__device__
--------------------------------------------------------------------------------

The ``__device__`` qualifier is used to specify functions for execution on the
device. They can only be called from other ``__device__`` functions or from
``__global__`` functions.

You can combine it with the ``__host__`` qualifier and mark functions
``__host__ __device__``. In this case, the function is compiled for the host and
the device. Note that these functions can't use the HIP built-ins (e.g.,

> ``threadIdx.x`` or ``warpSize``), as they are not available on the host. If you

Reviewer: Add link to threadIdx and warpSize

the device. Note that these functions can't use the HIP built-ins (e.g.,
:ref:`threadIdx.x <thread_and_block_idx>` or :ref:`warpSize <warp_size>`), as
they are not available on the host. If you need to use HIP grid coordinate
functions, you can pass the necessary coordinate information as an argument.

__global__
--------------------------------------------------------------------------------

Functions marked ``__global__`` are executed on the device and are referred to
as kernels. Their return type must be ``void``. Kernels have a special launch
mechanism, and have to be launched from the host.

There are some restrictions on the parameters of kernels. Kernels can't:

* have a parameter of type ``std::initializer_list`` or ``va_list``
* have a variable number of arguments
* use references as parameters
* use parameters having different sizes in host and device code, e.g. long double arguments, or structs containing long double members.
* use struct-type arguments which have different layouts in host and device code.

Kernels can have variadic template parameters, but only one parameter pack,
which must be the last item in the template parameter list.

> Unlike CUDA, HIP does not support dynamic parallelism, meaning that kernels can

Reviewer 2: I think this should be a Note or Important or something?

Author: Changed it to a note. I was thinking about removing it, as the first paragraph already states that they "have to be launched from the host", but the difference to CUDA is big enough that it warrants a mention.

not be called from the device.

Calling __global__ functions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The launch mechanism for kernels differs from standard function calls, as they
need an additional configuration, that specifies the grid and block dimensions
(i.e. the amount of threads to be launched), as well as specifying the amount of
shared memory per block and which stream to execute the kernel on.

> Kernels are called using the ``<<<>>>`` syntax known from CUDA, but HIP also

Reviewer 2: ```suggestion
Kernels are called using the triple chevron ``<<<>>>`` syntax known from CUDA, but HIP also
```

supports the ``hipLaunchKernelGGL`` macro.

> When using ``hipLaunchKernelGGL``, the first five configuration parameters must

Reviewer 2: Is there a Reference link for hipLaunchKernelGGL?

Author: Sadly no, as it is a macro. Its use has actually been discouraged, but since it is already so widespread I chose to keep it in here.

GPU multiprocessors have a fixed pool of resources (primarily registers and
shared memory) which are shared by the actively running warps. Using more
resources can increase IPC of the kernel but reduces the resources available for
other warps and limits the number of warps that can be simultaneously running.
Thus GPUs have to balance resource usage between instruction- and thread-level
parallelism.

``__launch_bounds__`` allows the application to provide hints that influence the
resource (primarily registers) usage of the generated code. It is a function
attribute that must be attached to a __global__ function:

.. code-block:: cpp

  __global__ void __launch_bounds__(MAX_THREADS_PER_BLOCK, MIN_WARPS_PER_EXECUTION_UNIT)
  kernel_name(/*args*/);

The  ``__launch_bounds__`` parameters are explained in the following sections:

MAX_THREADS_PER_BLOCK
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This parameter is a guarantee from the programmer, that kernel will not be
launched with more threads than ``MAX_THREADS_PER_BLOCK``. 

If no ``__launch_bounds__`` are specified, ``MAX_THREADS_PER_BLOCK`` is
the maximum block size supported by the device (see

> :doc:`reference/hardware_features`). Reducing ``MAX_THREADS_PER_BLOCK`` allows

Reviewer 2: ```suggestion
:doc:`../reference/hardware_features`). Reducing ``MAX_THREADS_PER_BLOCK`` allows
```

the compiler to use more resources per thread than an unconstrained
compilation. This might however reduce the amount of blocks that can run
concurrently on a CU, thereby reducing occupancy and trading thread-level
parallelism for instruction-level parallelism.

``MAX_THREADS_PER_BLOCK`` is particularly useful in cases, where the compiler is
constrained by register usage in order to meet requirements of large block sizes
that are never used at launch time.

The compiler can only use the hints to manage register usage, and does not
automatically reduce shared memory usage. The compilation fails, if the compiler
can not generate code that satisfies the launch bounds.

> On NVCC this parameter maps to the ``.maxntid`` PTX directive.

Reviewer 2: ```suggestion
On NVCC this parameter maps to the ``.maxntid`` PTX directive.
```
Is this needed?

Author: Yes and no. This and the "porting from CUDA __launch_bounds" section should go to the porting guide, but that is being done in another issue. I'll add a note to that issue

When launching kernels HIP will validate the launch configuration to make sure
the requested block size is not larger than ``MAX_THREADS_PER_BLOCK`` and
return an error if it is exceeded.

> If :doc:`AMD_LOG_LEVEL <how-to/logging>` is set, detailed information will be

Reviewer 2: ```suggestion
If :doc:`AMD_LOG_LEVEL <./logging>` is set, detailed information will be
```

kernel and the specified ``__launch_bounds__``.

MIN_WARPS_PER_EXECUTION_UNIT
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This parameter specifies the minimum number of warps that must be able to run
concurrently on an execution unit.
``MIN_WARPS_PER_EXECUTION_UNIT`` is optional and defaults to 1 if not specified.
Since active warps compete for the same fixed pool of resources, the compiler
must constrain the resource usage of the warps. This option gives a lower
bound to the occupancy of the kernel.

From this parameter, the compiler derives a maximum number of registers that can
be used in the kernel. The amount of registers that can be used at most is
:math:`\frac{\text{available registers}}{\text{MIN_WARPS_PER_EXECUTION_UNIT}}`,
but it might also have other, architecture specific, restrictions.

The available registers per Compute Unit are listed in
:doc:`rocm:reference/gpu-arch-specs`. Beware that these values are per Compute
Unit, not per Execution Unit. On AMD GPUs a Compute Unit consists of 4 Execution
Units, also known as SIMDs, each with their own register file. For more

> information see :doc:`understand/hardware_implementation`.

Reviewer 2: ```suggestion
information see :doc:`../understand/hardware_implementation`.
```

:cpp:struct:`hipDeviceProp_t` also has a field ``executionUnitsPerMultiprocessor``.

> Porting from CUDA __launch_bounds

Reviewer 2: ```suggestion
Porting CUDA __launch_bounds to HIP
```

Author: I think it's somewhat obvious that this is porting "to HIP", as this is the HIP documentation

> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Reviewer 2: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

CUDA defines the ``__launch_bounds`` qualifier which works similar to
``__launch_bounds__``:

.. code-block:: cpp

  __launch_bounds(MAX_THREADS_PER_BLOCK, MIN_BLOCKS_PER_MULTIPROCESSOR)

> The first parameter is the same as HIP's implementation. The second parameter of

Reviewer 2: ```suggestion
The first parameter is the same as HIP's implementation. The second parameter, ``MIN_BLOCKS_PER_MULTIPROCESSOR``,
```

Author: Reworded the whole phrase

> ``__launch_bounds`` must  be converted to the format used ``__launch_bounds__``,

Reviewer 2: ```suggestion
 must  be converted to ``MIN_WARPS_PER_EXECUTION`` as needed for HIP,
```

which uses warps and execution units rather than blocks and multiprocessors.
This conversion is performed automatically by :doc:`HIPIFY <hipify:index>`.

.. code-block:: cpp

  MIN_WARPS_PER_EXECUTION_UNIT = (MIN_BLOCKS_PER_MULTIPROCESSOR * MAX_THREADS_PER_BLOCK) / warpSize

Directly controlling the warps per execution unit makes it easier to reason
about the occupancy, unlike with blocks, where the occupancy depends on the
block size.

The use of execution units rather than multiprocessors also provides support for
architectures with multiple execution units per multiprocessor. For example, the
AMD GCN architecture has 4 execution units per multiprocessor.

maxregcount
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

> Unlike ``nvcc``, ``amdclang++`` does not support the ``--maxregcount`` option.

Reviewer 2: ```suggestion
``amdclang++`` does not support the ``--maxregcount`` option.
```

Author: Granted, this makes the sentence sound somewhat opinionated, but if not comparing it to nvcc, then why even mention it? There are a myriad of options out there that amdclang does not support, but we don't mention it because they are irrelevant. People coming from CUDA/nvcc however have an expectation towards HIP/amdclang++

Reviewer 2: The fact that an option makes is not supported, but an alternate launch mechanism provides the same feature might be worth mentioning, even without 'nvcc'

Author: Good point, but I'll leave it in for now, as it is also included in the issue for refactoring the porting guide

Instead, users are encouraged to use the ``__launch_bounds__`` directive since
the parameters are more intuitive and portable than micro-architecture details
like registers. The directive allows per-kernel control.

> ``__launch_bounds__`` works on both AMD and NVIDIA platforms.

Reviewer 2: ```suggestion
```

it perfectly suited for sharing variables.

Shared memory can be allocated statically within the kernel, but the size
of it has to be known at compile time.

In order to dynamically allocate shared memory during runtime, but before the
kernel is launched, the variable has to be declared  ``extern``, and the kernel
launch has to specify the needed amount of ``extern`` shared memory in the launch
configuration. The statically allocated shared memory is allocated without this
parameter.

.. code-block:: cpp

  #include <hip/hip_runtime.h>

  extern __shared__ int shared_array[];

  __global__ void kernel(){
    // initialize shared memory
    shared_array[threadIdx.x] = threadIdx.x;
    // use shared memory
  }

  int main(){
    //shared memory in this case depends on the configurable block size
    constexpr int blockSize = 256;
    constexpr int sharedMemSize = blockSize * sizeof(int);
    constexpr int gridSize = 2;

    kernel<<<dim3(gridSize), dim3(blockSize), sharedMemSize, 0>>>();
  }

__managed__
--------------------------------------------------------------------------------

Managed memory is a special qualifier, that makes the marked memory available on

> the device and on the host. For more details see :ref:`unified_memory`.

Reviewer 2: ```suggestion
the device and on the host. For more details see :ref:`unified_memory`, or :doc:`../reference/hip_runtime_api/modules/memory_management/unified_memory_reference`.
```

Reviewer 2: Just seems like this reference could be more than just the Unified Memory topic

Author: I think the reference that just lists the functions without explaining the concept is not of much use here. Especially, as the reference for unified memory does not mention the __managed__ qualifier. The reference is however linked in the already linked page

HIP supports vectors created from the following fundamental types:

.. list-table::
  *
    - :cspan:`1` Integral Types
  *
    - ``char``
    - ``uchar``
  *
    - ``short``
    - ``ushort``
  *
    - ``int``
    - ``uint``
  *
    - ``long``
    - ``ulong``
  *
    - ``longlong``
    - ``ulonglong``
  *
    - :cspan:`1` Floating-Point Types
  *
    - :cspan:`1` ``float``
  *
    - :cspan:`1` ``double``

.. _dim3:

dim3
================================================================================

``dim3`` is a special three-dimensional unsigned integer vector type that is
commonly used to specify grid and group dimensions for kernel launch
configurations.

Its constructor accepts up to three arguments. The unspecified dimensions are
initialized to 1.

********************************************************************************
Built-in device functions
********************************************************************************

.. _memory_fence_instructions:

Memory fence instructions
================================================================================

HIP supports ``__threadfence()``, ``__threadfence_block()`` and
``__threadfence_system()``.

> On AMD devices, ``__threadfence_system()``, has restrictions and therefore needs

Reviewer 2: This feels like it is referring to some implicit knowledge the user must have relative to threadfence_system? What are the "restrictions"? And is it appropriate to call it a workaround in the docs? Seems like a workaround is good for an issue or bug report, but not for user docs. Perhaps "requires the following process"?

Author: I wish I had more information on this, but I couldn't find information proving or disproving this, so I just reformatted the already existing content

Reviewer 2: requires the following process?

Author: I'll ask around what the current status on this feature is, maybe we can get rid of this section altogether, but I'll postpone it for now

(i.e. the amount of threads to be launched), as well as specifying the amount of
shared memory per block and which stream to execute the kernel on.

Kernels are called using the triple chevron ``<<<>>>`` syntax known from CUDA,
but HIP also supports the ``hipLaunchKernelGGL`` macro.

When using ``hipLaunchKernelGGL``, the first five configuration parameters must
be:

* ``symbol kernelName``: The name of the kernel you want to launch. To support
  template kernels that contain several template parameters separated by use the
  ``HIP_KERNEL_NAME`` macro to wrap the template instantiation
  (:doc:`HIPIFY <hipify:index>` inserts this automatically).
* ``dim3 gridDim``: 3D-grid dimensions that specifies the number of blocks to
  launch.
* ``dim3 blockDim``: 3D-block dimensions that specifies the number of threads in
  each block.
* ``size_t dynamicShared``: The amount of additional shared dynamic memory to
  allocate per block.
* ``hipStream_t``: The stream on which to run the kernel. A value of ``0``
  corresponds to the default stream.

The kernel arguments are listed after the configuration parameters.

.. code-block:: cpp

>   #include <hip/hip_runtime.h>

Reviewer 3: can we modify this to add __host__ and __device__ functions as well?

Author: This example is specifically about launching a kernel and I'd like to keep it as concise as possible

Reviewer 3: Make sense, but this kind of felt like a build up, there is description of host, device and global functions but example only shows global function. I think user reading through it, when they reach this section should get a complete picture of all three in action.

I'll leave it up to you, but if I was reading this, I wish there would be a more complete example which encompasses all that I just read.

Author: I added a `__device__` and a `__host__ __device__` function. At least it shows how they are defined and used.

  __global__ void example_kernel(float * const a, const unsigned int N)
  {
    // Index variables. Determined by the launch configuration.
    // The following uniquely identifies a thread in a 1D configuration.
    const int globalIdx = threadIdx.x + blockIdx.x * blockDim.x;
    // simple initialization of the array
    if(globalIdx < N){
      a[globalIdx] = globalIdx;
    }
  }

  int main()
  {
    constexpr int N = 1000000; // problem size
    constexpr int blockSize = 256; //configurable block size
    constexpr int gridSize = (N + blockSize - 1)/blockSize; //needed number of blocks for the given problem size

    float *a;
    hipMalloc(&a, sizeof(*a) * N);

>     example_kernel<<<dim3(gridSize), dim3(blockSize), 0/*example doesn't use shared memory*/, 0/*default stream*/>>>(a, N);

Reviewer 3: If a user compiles this, they might get a warning, of discarding the no-discard attribute of hipMalloc. Can we handle the return type and also add corresponding hipFree?

Author: Good point. Updated the example code

  }

Inline qualifiers
--------------------------------------------------------------------------------

HIP adds the ``__noinline__`` and ``__forceinline__`` function qualifiers.

``__noinline__`` is a hint to the compiler to not inline the function, whereas 
``__forceinline__`` forces the compiler to inline the function. These qualifiers

> can be applied to both ``__host__`` and ``__device__`` functions.

Reviewer 3: AFAIK, __device__ functions will be inlined, even if you add noinline to it.

Author: That might be true for compilation for AMD devices, however when targeting e.g. CUDA, this is not "guaranteed", there it is just a hint

``__noinline__`` and ``__forceinline__`` can not be used in combination.

__launch_bounds__
--------------------------------------------------------------------------------

GPU multiprocessors have a fixed pool of resources (primarily registers and
shared memory) which are shared by the actively running warps. Using more

> resources can increase IPC of the kernel but reduces the resources available for

Reviewer 3: > can increase IPC of the kernel

What does IPC mean here?

Author: instructions per cycle. Rephrased the sentence to replace the acronym

Reviewer 3: Should we use such words (IPC incase of kernels?). Can we measure IPC for kernels? I would recommend using something like 'occupancy' which is a measurable metric.

Author: Yes, IPC is very much measurable. Occupancy was already implicitly included in that sentence ("limits the number of warps that can be simultaneously running"), but I rephrased it to explicitly mention occupancy.

MAX_THREADS_PER_BLOCK
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This parameter is a guarantee from the programmer, that kernel will not be
launched with more threads than ``MAX_THREADS_PER_BLOCK``. 

If no ``__launch_bounds__`` are specified, ``MAX_THREADS_PER_BLOCK`` is
the maximum block size supported by the device (see
:doc:`../reference/hardware_features`). Reducing ``MAX_THREADS_PER_BLOCK``
allows the compiler to use more resources per thread than an unconstrained
compilation. This might however reduce the amount of blocks that can run
concurrently on a CU, thereby reducing occupancy and trading thread-level
parallelism for instruction-level parallelism.

``MAX_THREADS_PER_BLOCK`` is particularly useful in cases, where the compiler is
constrained by register usage in order to meet requirements of large block sizes
that are never used at launch time.

The compiler can only use the hints to manage register usage, and does not
automatically reduce shared memory usage. The compilation fails, if the compiler
can not generate code that satisfies the launch bounds.

> On NVCC this parameter maps to the ``.maxntid`` PTX directive.

Reviewer 3: If we are mentioning PTX directive, should we also mention the `amdgpu_flat_work_group_size` function attribute?

Author: I agree, but this also is not the appropriate place in general. I already opened an issue to move this to the compiler documentation, as this does not directly relate to the language

concurrently on an execution unit.
``MIN_WARPS_PER_EXECUTION_UNIT`` is optional and defaults to 1 if not specified.
Since active warps compete for the same fixed pool of resources, the compiler
must constrain the resource usage of the warps. This option gives a lower
bound to the occupancy of the kernel.

From this parameter, the compiler derives a maximum number of registers that can
be used in the kernel. The amount of registers that can be used at most is
:math:`\frac{\text{available registers}}{\text{MIN_WARPS_PER_EXECUTION_UNIT}}`,
but it might also have other, architecture specific, restrictions.

The available registers per Compute Unit are listed in
:doc:`rocm:reference/gpu-arch-specs`. Beware that these values are per Compute
Unit, not per Execution Unit. On AMD GPUs a Compute Unit consists of 4 Execution
Units, also known as SIMDs, each with their own register file. For more
information see :doc:`../understand/hardware_implementation`.
:cpp:struct:`hipDeviceProp_t` also has a field ``executionUnitsPerMultiprocessor``.

Porting from CUDA __launch_bounds
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> CUDA defines the ``__launch_bounds`` qualifier which works similar to

Reviewer 3: Does CUDA support __launch_bounds (without the trailing underscore?) can you point me to the CUDA documentation for it, cant  seem to find it.

Author: I could have sworn I've seen it somewhere, but it seems you're right, it also has trailing underscores. Fixed it.

    void set_constant_memory(){
      int host_data[8] {1,2,3,4,5,6,7,8};

      hipMemcpyToSymbol(const_array, host_data, sizeof(int) * 8);

      // call kernel that accesses const_array
    }

__shared__
--------------------------------------------------------------------------------

Variables marked with ``__shared__`` are only accessible by threads within the
same block and have the lifetime of that block. It is usually backed by on-chip
shared memory, providing fast access to all threads within a block, which makes
it perfectly suited for sharing variables.

Shared memory can be allocated statically within the kernel, but the size
of it has to be known at compile time.

In order to dynamically allocate shared memory during runtime, but before the
kernel is launched, the variable has to be declared  ``extern``, and the kernel
launch has to specify the needed amount of ``extern`` shared memory in the launch
configuration. The statically allocated shared memory is allocated without this
parameter.

.. code-block:: cpp

  #include <hip/hip_runtime.h>

  extern __shared__ int shared_array[];

  __global__ void kernel(){
    // initialize shared memory

>     shared_array[threadIdx.x] = threadIdx.x;

Reviewer 3: should we also mention a small note of __syncthreads here?

Author: Good point. Added a note and a __syncthreads invocation.

    // use shared memory
  }

  int main(){
    //shared memory in this case depends on the configurable block size
    constexpr int blockSize = 256;
    constexpr int sharedMemSize = blockSize * sizeof(int);
    constexpr int gridSize = 2;

    kernel<<<dim3(gridSize), dim3(blockSize), sharedMemSize, 0>>>();
  }

__managed__
--------------------------------------------------------------------------------

Managed memory is a special qualifier, that makes the marked memory available on
the device and on the host. For more details see :ref:`unified_memory`.

> __restrict__

Reviewer 3: Isnt restrict more of a clang extension than HIP, should we include it in HIP documentation? or alteast mention that here?

Author: It is somewhat of a language extension, but also compiler specific/a hint to the compiler. However since CUDA requires this, I'm quite sure HIP also has to support it, so I'd say it has its place in this documentation
